modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+    // an actual lock because it is accessed only from the main thread.
+    bool m_GenKeyRunning;
+    void* m_NewKey;
+    gcry_error_t m_GenKeyError;
+
+  public:
+    MODCONSTRUCTOR(COtrMod) {}
+
+    bool PutModuleBuffered(const CString& sLine) {
+        CUser* user = GetUser();
+        bool attached = user->IsUserAttached();
+        if (attached) {
+            PutModule(sLine);
+        } else {
+            m_Buffer.push_back(sLine);
+        }
+        return attached;
+    }
+
+    bool PutModuleContext(ConnContext* ctx, const CString& sLine) {
+        assert(ctx);
+        assert(ctx->username);
+        return PutModuleBuffered(CString("[") + ctx->username + "] " + sLine);
+    }
+
+    enum Color { Blue = 2, Green = 3, Red = 4, Bold = 32 };
+
+    CString Clr(Color eClr, const CString& sWhat) {
+        if (eClr == Bold) {
+            return CString("\x02") + sWhat + CString("\x02");
+        } else {
+            return CString("\x03") + CString(eClr) + sWhat + CString("\x03");
+        }
+    }
+
+    static CString HumanFingerprint(Fingerprint* fprint) {
+        char human[OTRL_PRIVKEY_FPRINT_HUMAN_LEN];
+        otrl_privkey_hash_to_human(human, fprint->fingerprint);
+        return CString(human);
+    }
+
+    CString OurFingerprint() {
+        char ourfp[OTRL_PRIVKEY_FPRINT_HUMAN_LEN];
+        const char* accountname = GetUser()->GetUserName().c_str();
+        if (otrl_privkey_fingerprint(m_pUserState, ourfp, accountname,
+                                     PROTOCOL_ID)) {
+            return CString(ourfp);
+        }
+        return CString("ERROR");
+    }
+
+    void WriteFingerprints() {
+        gcry_error_t err;
+        err = otrl_privkey_write_fingerprints(m_pUserState, m_sFPPath.c_str());
+        if (err) {
+            PutModuleBuffered(CString("Failed to write fingerprints: ") +
+                              gcry_strerror(err));
+        }
+    }
+
+    void WarnIfLoggingEnabled() {
+        assert(GetNetwork());
+        bool bUserMod = GetUser()->GetModules().FindModule("log");
+        bool bNetworkMod = GetNetwork()->GetModules().FindModule("log");
+
+        if (bUserMod || bNetworkMod) {
+            CString sMsg =
+                Clr(Red, "WARNING:") + " The log module is loaded. Type ";
+            if (bUserMod) {
+                sMsg += Clr(Bold, "/msg *status UnloadMod log") + " ";
+                if (bNetworkMod) {
+                    sMsg += "and ";
+                }
+            }
+            if (bNetworkMod) {
+                sMsg += Clr(Bold, "/msg *status UnloadMod --type=network log") +
+                        " ";
+            }
+            sMsg += "to prevent ZNC from logging the conversation to disk.";
+            PutModuleBuffered(sMsg);
+        }
+    }
+
+    void CmdInfo(const CString& sLine) {
+        CTable table;
+        table.AddColumn("Peer");
+        table.AddColumn("State");
+        table.AddColumn("Fingerprint");
+        table.AddColumn("Act");
```

Comment:
What is act? acting?
Reading code below, it's "active", but user probably doesn't read the code